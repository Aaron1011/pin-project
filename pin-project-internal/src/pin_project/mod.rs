use proc_macro2::{Ident, Span, TokenStream};
use quote::{quote, quote_spanned};
use syn::{
    parse::{Parse, ParseStream},
    Fields, FieldsNamed, FieldsUnnamed, Generics, Index, Item, ItemStruct, Meta, NestedMeta,
    Result, Type,
};

use crate::utils::crate_path;

mod enums;
mod structs;

/// The annotation for pinned type.
const PIN: &str = "pin";

pub(super) fn attribute(args: TokenStream, input: TokenStream) -> TokenStream {
    parse(args, input).unwrap_or_else(|e| e.to_compile_error())
}

#[derive(Clone, Copy)]
struct Args {
    pinned_drop: Option<Span>,
    unsafe_unpin: Option<Span>,
}

impl Args {
    fn impl_drop(self, generics: &Generics) -> ImplDrop<'_> {
        ImplDrop::new(generics, self.pinned_drop)
    }

    fn impl_unpin(self, generics: &Generics) -> ImplUnpin {
        ImplUnpin::new(generics, self.unsafe_unpin)
    }
}

impl Parse for Args {
    fn parse(input: ParseStream<'_>) -> Result<Self> {
        let mut pinned_drop = None;
        let mut unsafe_unpin = None;
        while !input.is_empty() {
            let i = input.parse::<Ident>()?;
            match &*i.to_string() {
                "PinnedDrop" => pinned_drop = Some(i.span()),
                "UnsafeUnpin" => unsafe_unpin = Some(i.span()),
                _ => return Err(error!(i, "an invalid argument was passed")),
            }
        }
        Ok(Self { pinned_drop, unsafe_unpin })
    }
}

fn parse(args: TokenStream, input: TokenStream) -> Result<TokenStream> {
    let args = syn::parse2(args)?;
    match syn::parse2(input)? {
        Item::Struct(item) => {
            let packed_check = ensure_not_packed(&item)?;
            let mut res = structs::parse(args, item)?;
            res.extend(packed_check);
            Ok(res)
        }
        Item::Enum(item) => {
            // We don't need to check for '#[repr(packed)]',
            // since it does not apply to enums
            enums::parse(args, item)
        }
        item => Err(error!(item, "may only be used on structs or enums")),
    }
}

fn ensure_not_packed(item: &ItemStruct) -> Result<TokenStream> {
    for meta in item.attrs.iter().filter_map(|attr| attr.parse_meta().ok()) {
        if let Meta::List(l) = meta {
            if l.ident == "repr" {
                for repr in l.nested.iter() {
                    if let NestedMeta::Meta(Meta::Word(w)) = repr {
                        if w == "packed" {
                            return Err(error!(
                                w,
                                "pin_project may not be used on #[repr(packed)] types"
                            ));
                        }
                    }
                }
            }
        }
    }

    // Workaround for https://github.com/taiki-e/pin-project/issues/32
    // Through the tricky use of proc macros, it's possible to bypass
    // the above check for the 'repr' attribute.
    // To ensure that it's impossible to use pin projections on a #[repr(packed)][
    // struct, we generate code like this:
    //
    // #[deny(safe_packed_borrows)]
    // fn enforce_not_packed_for_MyStruct(val: MyStruct) {
    //  let _field1 = &val.field1;
    //  let _field2 = &val.field2;
    //  ...
    //  let _fieldn = &val.fieldn;
    // }
    //
    // Taking a reference to a packed field is unsafe, amd appplying
    // #[deny(safe_packed_borrows)] makes sure that doing this without
    // an 'unsafe' block (which we deliberately do not generate)
    // is a hard error.
    //
    // If the struct ends up having #[repr(packed)] applied somehow,
    // this will generate an (unfriendly) error message. Under all reasonable
    // circumstances, we'll detect the #[repr(packed)] attribute, and generate
    // a much nicer error above.
    //
    // There is one exception: If the type of a struct field has a alignemtn of 1
    // (e.g. u8), it is always safe to take a reference to it, even if the struct
    // is #[repr(packed)]. If the struct is composed entirely of types of alignent 1,
    // our generated method will not trigger an error if the struct is #[repr(packed)]
    //
    // Fortunately, this should have no observable consequence - #[repr(packed)]
    // is essentially a no-op on such a type. Nevertheless, we include a test
    // to ensure that the compiler doesn't ever try to copy the fields on
    // such a struct when trying to drop it - which is reason we prevent
    // #[repr(packed)] in the first place
    let mut field_refs = vec![];
    match &item.fields {
        Fields::Named(FieldsNamed { named, .. }) => {
            for field in named.iter() {
                let ident = field.ident.as_ref().unwrap();
                field_refs.push(quote!(&val.#ident;));
            }
        }
        Fields::Unnamed(FieldsUnnamed { unnamed, .. }) => {
            for (i, _) in unnamed.iter().enumerate() {
                let index = Index::from(i);
                field_refs.push(quote!(&val.#index;));
            }
        }
        Fields::Unit => {}
    }

    let (impl_generics, ty_generics, where_clause) = item.generics.split_for_impl();

    let struct_name = &item.ident;
    let method_name = Ident::new(
        &("__pin_project_assert_not_repr_packed_".to_string() + &item.ident.to_string()),
        Span::call_site(),
    );
    let test_fn = quote! {
        #[deny(safe_packed_borrows)]
        fn #method_name #impl_generics (val: #struct_name #ty_generics) #where_clause {
            #(#field_refs)*
        }
    };
    Ok(test_fn)
}

/// Makes the generics of projected type from the reference of the original generics.
fn proj_generics(generics: &Generics) -> Generics {
    let mut generics = generics.clone();
    generics.params.insert(0, syn::parse_quote!('__a));
    generics
}

// =================================================================================================
// Drop implementation

struct ImplDrop<'a> {
    generics: &'a Generics,
    pinned_drop: Option<Span>,
}

impl<'a> ImplDrop<'a> {
    fn new(generics: &'a Generics, pinned_drop: Option<Span>) -> Self {
        Self { generics, pinned_drop }
    }

    fn build(self, ident: &Ident) -> TokenStream {
        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();

        if let Some(pinned_drop) = self.pinned_drop {
            let crate_path = crate_path();
            let call = quote_spanned! { pinned_drop =>
                ::#crate_path::__private::UnsafePinnedDrop::pinned_drop(pinned_self)
            };

            quote! {
                impl #impl_generics ::core::ops::Drop for #ident #ty_generics #where_clause {
                    fn drop(&mut self) {
                        // Safety - we're in 'drop', so we know that 'self' will
                        // never move again
                        let pinned_self = unsafe { ::core::pin::Pin::new_unchecked(self) };
                        // We call `pinned_drop` only once. Since `UnsafePinnedDrop::pinned_drop`
                        // is an unsafe function and a private API, it is never called again in safe
                        // code *unless the user uses a maliciously crafted macro*.
                        unsafe {
                            #call;
                        }
                    }
                }
            }
        } else {
            quote! {
                impl #impl_generics ::core::ops::Drop for #ident #ty_generics #where_clause {
                    fn drop(&mut self) {
                        // Do nothing. The precense of this Drop
                        // impl ensures that the user can't provide one of their own
                    }
                }
            }
        }
    }
}

// =================================================================================================
// conditional Unpin implementation

struct ImplUnpin {
    generics: Generics,
    unsafe_unpin: bool,
}

impl ImplUnpin {
    fn new(generics: &Generics, unsafe_unpin: Option<Span>) -> Self {
        let mut generics = generics.clone();
        if let Some(unsafe_unpin) = unsafe_unpin {
            let crate_path = crate_path();
            generics.make_where_clause().predicates.push(
                syn::parse2(quote_spanned! { unsafe_unpin =>
                    ::#crate_path::__private::Wrapper<Self>: ::#crate_path::UnsafeUnpin
                })
                .unwrap(),
            );
        }

        Self { generics, unsafe_unpin: unsafe_unpin.is_some() }
    }

    fn push(&mut self, ty: &Type) {
        // We only add bounds for automatically generated impls
        if !self.unsafe_unpin {
            self.generics
                .make_where_clause()
                .predicates
                .push(syn::parse_quote!(#ty: ::core::marker::Unpin));
        }
    }

    /// Creates `Unpin` implementation.
    fn build(self, ident: &Ident) -> TokenStream {
        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();
        quote! {
            impl #impl_generics ::core::marker::Unpin for #ident #ty_generics #where_clause {}
        }
    }
}
