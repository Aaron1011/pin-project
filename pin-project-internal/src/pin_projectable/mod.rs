use proc_macro2::{Ident, Span, TokenStream};
use quote::{quote, ToTokens};
use syn::parse::{Parse, ParseStream};
use syn::{
    Fields, FieldsNamed, FieldsUnnamed, Generics, Item, ItemEnum, ItemFn, ItemStruct, Meta,
    NestedMeta, Result, ReturnType, Type, TypeTuple, Index
};

use crate::utils::{Nothing, VecExt};

mod enums;
mod structs;

/// The annotation for pinned type.
const PIN: &str = "pin";

const PINNED_DROP: &str = "pinned_drop";

struct PinProject {
    items: Vec<Item>,
}

impl Parse for PinProject {
    fn parse(input: ParseStream<'_>) -> Result<Self> {
        let mut items = vec![];
        while !input.is_empty() {
            items.push(input.parse()?);
        }
        Ok(Self { items })
    }
}

fn handle_type(args: TokenStream, item: Item, pinned_drop: Option<ItemFn>) -> Result<TokenStream> {
    match item {
        Item::Struct(item) => {
            let packed_check = ensure_not_packed(&item)?;
            let mut res = structs::parse(args, item, pinned_drop)?;
            //println!("Packed check: {}", packed_check);
            res.extend(packed_check);
            Ok(res)
        }
        Item::Enum(item) => {
            // We don't need to check for '#[repr(packed)]',
            // since it does not apply to enums
            enums::parse(args, item, pinned_drop)
        }
        _ => unreachable!(),
    }
}

pub(super) fn pin_project(input: TokenStream) -> Result<TokenStream> {
    let span = span!(input);
    let items: Vec<Item> = syn::parse2::<PinProject>(input)?.items;

    let mut found_type = None;
    let mut found_pinned_drop = None;

    for mut item in items {
        match &mut item {
            Item::Struct(ItemStruct { attrs, .. }) | Item::Enum(ItemEnum { attrs, .. }) => {
                if found_type.is_none() {
                    if let Some(attr) = attrs.find_remove("pin_projectable") {
                        let args = match attr.parse_meta()? {
                            Meta::List(l) => l.nested.into_token_stream(),
                            Meta::Word(_) => TokenStream::new(),
                            _ => return Err(error!(span!(attr), "invalid arguments"))
                        };

                        found_type = Some((item.clone(), args));
                    } else {
                        return Err(error!(item, "type declared in pin_project! must have pin_projectable attribute"))
                    }
                } else {
                    return Err(error!(item, "cannot declare multiple types within pinned module"))
                }
            },
            Item::Fn(fn_) => {
                if let Some(attr)= fn_.attrs.find_remove(PINNED_DROP) {
                    let _: Nothing = syn::parse2(attr.tts)?;
                    if found_pinned_drop.is_none() {
                        if let ReturnType::Type(_, ty) = &fn_.decl.output {
                            match &**ty {
                                Type::Tuple(TypeTuple { elems, .. }) if elems.is_empty() => {}
                                _ => return Err(error!(ty, "#[pinned_drop] function must return the unit type")),
                            }
                        }
                        found_pinned_drop = Some(fn_.clone());
                    } else {
                        return Err(error!(fn_, "cannot declare multiple functions within pinned module"));
                    }
                } else {
                    return Err(error!(fn_, "only #[pinned_drop] functions can be declared within a pin_project! block"));
                }
            },
            _ => return Err(error!(item, "pin_project! block may only contain a struct/enum with an optional #[pinned_drop] function"))
        }
    }

    let (type_, args) = match found_type {
        Some(t) => t,
        None => return Err(error!(span, "No #[pin_projectable] type found!")),
    };

    handle_type(args, type_, found_pinned_drop)
}

pub(super) fn attribute(args: TokenStream, input: TokenStream) -> Result<TokenStream> {
    let item = syn::parse2(input)?;
    match &item {
        Item::Struct(_) | Item::Enum(_) => handle_type(args, item, None),
        _ => Err(error!(item, "may only be used on structs or enums")),
    }
}

fn ensure_not_packed(item: &ItemStruct) -> Result<TokenStream> {
    for meta in item.attrs.iter().filter_map(|attr| attr.parse_meta().ok()) {
        if let Meta::List(l) = meta {
            if l.ident == "repr" {
                for repr in l.nested.iter() {
                    if let NestedMeta::Meta(Meta::Word(w)) = repr {
                        if w == "packed" {
                            return Err(error!(
                                w,
                                "pin_projectable may not be used on #[repr(packed)] types"
                            ));
                        }
                    }
                }
            }
        }
    }

    // Workaround for https://github.com/taiki-e/pin-project/issues/32
    // Through the tricky use of proc macros, it's possible to bypass
    // the above check for the 'repr' attribute.
    // To ensure that it's impossible to use pin projections on a #[repr(packed)][
    // struct, we generate code like this:
    //
    // #[deny(safe_packed_borrows)]
    // fn enforce_not_packed_for_MyStruct(val: MyStruct) {
    //  let _field1 = &val.field1;
    //  let _field2 = &val.field2;
    //  ...
    //  let _fieldn = &val.fieldn;
    // }
    //
    // Taking a reference to a packed field is unsafe, amd appplying
    // #[deny(safe_packed_borrows)] makes sure that doing this without
    // an 'unsafe' block (which we deliberately do not generate)
    // is a hard error.
    //
    // If the struct ends up having #[repr(packed)] applied somehow,
    // this will generate an (unfriendly) error message. Under all reasonable
    // circumstances, we'll detect the #[repr(packed)] attribute, and generate
    // a much nicer error above.
    //
    // There is one exception: If the type of a struct field has a alignemtn of 1
    // (e.g. u8), it is always safe to take a reference to it, even if the struct
    // is #[repr(packed)]. If the struct is composed entirely of types of alignent 1,
    // our generated method will not trigger an error if the struct is #[repr(packed)]
    //
    // Fortunately, this should have no observable consequence - #[repr(packed)]
    // is essentially a no-op on such a type. Nevertheless, we include a test
    // to ensure that the compiler doesn't ever try to copy the fields on
    // such a struct when trying to drop it - which is reason we prevent
    // #[repr(packed)] in the first place
    let mut field_refs = vec![];
    match &item.fields {
        Fields::Named(FieldsNamed { named, .. }) => {
            for field in named.iter() {
                let ident = field.ident.as_ref().unwrap();
                field_refs.push(quote!(&val.#ident;));
            }
        }
        Fields::Unnamed(FieldsUnnamed { unnamed, .. }) => {
            for (i, _) in unnamed.iter().enumerate() {
                let index = Index::from(i);
                field_refs.push(quote!(&val.#index;));
            }
        }
        Fields::Unit => {}
    }

    let (impl_generics, ty_generics, where_clause) = item.generics.split_for_impl();

    let struct_name = &item.ident;
    let method_name = Ident::new(
        &("__pin_project_assert_not_repr_packed_".to_string() + &item.ident.to_string()),
        Span::call_site(),
    );
    let test_fn = quote! {
        #[deny(safe_packed_borrows)]
        fn #method_name #impl_generics (val: #struct_name #ty_generics) #where_clause {
            #(#field_refs)*
        }
    };
    Ok(test_fn)
}

/// Makes the generics of projected type from the reference of the original generics.
fn proj_generics(generics: &Generics) -> Generics {
    let mut generics = generics.clone();
    generics.params.insert(0, syn::parse_quote!('__a));
    generics
}

struct ImplDrop {
    generics: Generics,
    pinned_drop: Option<ItemFn>,
}

impl ImplDrop {
    /// Parses attribute arguments.
    fn new(generics: Generics, pinned_drop: Option<ItemFn>) -> Result<Self> {
        Ok(Self { generics, pinned_drop })
    }

    fn build(self, ident: &Ident) -> TokenStream {
        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();

        if let Some(fn_) = self.pinned_drop {
            let fn_name = fn_.ident.clone();
            quote! {
                impl #impl_generics ::core::ops::Drop for #ident #ty_generics #where_clause {
                    fn drop(&mut self) {
                        // Declare the #[pinned_drop] function *inside* our drop function
                        // This guarantees that it's impossible for any other user code
                        // to call it
                        #fn_
                        // Safety - we're in 'drop', so we know that 'self' will
                        // never move again
                        let pinned_self = unsafe { ::core::pin::Pin::new_unchecked(self) };
                        // 'pinned_drop' is a free function - if it were part of a trait impl,
                        // it would be possible for user code to call it by directly invoking
                        // the trait.
                        #fn_name(pinned_self);
                    }
                }
            }
        } else {
            quote! {
                impl #impl_generics ::core::ops::Drop for #ident #ty_generics #where_clause {
                    fn drop(&mut self) {
                        // Do nothing. The precense of this Drop
                        // impl ensures that the user can't provide one of their own
                    }
                }
            }
        }
    }
}

// =================================================================================================
// conditional Unpin implementation

struct ImplUnpin {
    generics: Generics,
    auto: bool,
}

impl ImplUnpin {
    /// Parses attribute arguments.
    fn new(args: TokenStream, generics: &Generics) -> Result<Self> {
        let mut generics = generics.clone();
        generics.make_where_clause();

        match &*args.to_string() {
            "" => Ok(Self { generics, auto: true }),
            "unsafe_Unpin" => Ok(Self { generics, auto: false }),
            _ => Err(error!(args, "an invalid argument was passed")),
        }
    }

    fn push(&mut self, ty: &Type) {
        // We only add bounds for automatically generated impls
        if self.auto {
            self.generics
                .make_where_clause()
                .predicates
                .push(syn::parse_quote!(#ty: ::core::marker::Unpin));
        }
    }

    /// Creates `Unpin` implementation.
    fn build(self, ident: &Ident) -> TokenStream {
        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();
        let mut where_clause = where_clause.unwrap().clone(); // Created in 'new'

        if self.auto {
            quote! {
                impl #impl_generics ::core::marker::Unpin for #ident #ty_generics #where_clause {}
            }
        } else {
            let pin_project_crate = pin_project_crate_path();
            where_clause.predicates.push(syn::parse_quote!(::#pin_project_crate::Wrapper<#ident #ty_generics>: ::#pin_project_crate::UnsafeUnpin));

            quote! {
                impl #impl_generics ::core::marker::Unpin for #ident #ty_generics #where_clause {}
            }
        }
    }
}

/// If the 'renamed' feature is enabled, we detect
/// the actual name of the 'pin-project' crate in the consumer's Cargo.toml
#[cfg(feature = "renamed")]
fn pin_project_crate_path() -> Ident {
    use crate::PIN_PROJECT_CRATE;
    // This is fairly subtle.
    // Normally, you would use `env!("CARGO_PKG_NAME")` to get the name of the package,
    // since it's set at compile time.
    // However, we're in a proc macro, which runs while *another* crate is being compiled.
    // By retreiving the runtime value of `CARGO_PKG_NAME`, we can figure out the name
    // of the crate that's calling us.
    let cur_crate = std::env::var("CARGO_PKG_NAME")
        .expect("Could not find CARGO_PKG_NAME environemnt variable");
    Ident::new(
        if cur_crate == "pin-project" { "pin_project" } else { PIN_PROJECT_CRATE.as_str() },
        Span::call_site(),
    )
}

/// If the 'renamed' feature is not enabled, we just
/// assume that the 'pin-project' dependency has not been renamed
#[cfg(not(feature = "renamed"))]
fn pin_project_crate_path() -> Ident {
    Ident::new("pin_project", Span::call_site())
}
