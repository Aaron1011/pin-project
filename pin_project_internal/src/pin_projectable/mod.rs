use proc_macro2::{Ident, Span, TokenStream};
use quote::{quote_spanned, ToTokens};
use syn::{spanned::Spanned, Generics, Item, Result, Type, Attribute, Meta, NestedMeta, ItemFn, ItemStruct, ItemEnum, WhereClause};
use syn::parse::{Parse, ParseStream};
use syn::punctuated::Punctuated;

use crate::utils::VecExt;

mod enums;
mod structs;

/// The annotation for pinned type.
const PIN: &str = "pin";

const PINNED_DROP: &str = "pinned_drop";

struct PinProject {
    items: Vec<Item>
}

impl Parse for PinProject {
    fn parse(input: ParseStream) -> Result<Self> {
        let mut items = vec![];
        while !input.is_empty() {
            items.push(input.parse()?);
        }
        Ok(PinProject { items })
    }
}

fn handle_type(args: TokenStream, item: Item, pinned_drop: Option<ItemFn>) -> Result<TokenStream>  {
    match item {
        Item::Struct(item) => {
            ensure_not_packed(&item.attrs)?;
            Ok(structs::parse(args, item, pinned_drop)?)
        }
        Item::Enum(item) => {
            ensure_not_packed(&item.attrs)?;
            Ok(enums::parse(args, item, pinned_drop)?)
        },
        _ => panic!("Unexpected item: {:?}", item)
    }
}

pub(super) fn pin_project(input: TokenStream) -> Result<TokenStream> {
    let span = span!(input);
    let mut items: Vec<Item> = syn::parse2::<PinProject>(input)?.items;
    let mut found_type = None;
    let mut found_pinned_drop = None;

    for mut item in items {
        match &mut item {
            Item::Struct(ItemStruct { attrs, .. }) | Item::Enum(ItemEnum { attrs, .. }) => {
                if found_type.is_none() {
                    if let Some(pos) = attrs.iter().position(|a| a.path.is_ident("pin_projectable")) {
                        // Remove the 'pin_projectable' attribute, to prevent it from
                        // being parsed again
                        let attr = attrs.remove(pos);
                        let args = match attr.parse_meta()? {
                            Meta::List(l) => l.nested.into_token_stream(),
                            Meta::Word(_) => TokenStream::new(),
                            _ => return Err(error!(span!(attr), "invalid arguments"))
                        };

                        println!("Args: {:?}", args);
                        found_type = Some((item.clone(), args));
                    } else {
                        return Err(error!(span, "type delcared in pin_project! must have pin_projectable attribute"))
                    }
                } else {
                    return Err(error!(span, "cannot declare multiple types within pinned module"))
                }
            },
            Item::Fn(ref mut fn_) => {
                if fn_.attrs.find_remove(PINNED_DROP) {
                    if found_pinned_drop.is_none() {
                        found_pinned_drop = Some(fn_.clone());
                    } else {
                        return Err(error!(span, "cannot declare multiple functions within pinned module"));
                    }
                } else {
                    return Err(error!(span, "only #[pinned_drop] functions cannot be declared within pinend module"));
                }
            },
            _ => return Err(error!(span, "pinned module may only contain a struct/enum with an option #[pinned_drop] function"))
        }
    }

    if found_type.is_none() {
        return Err(error!(span, "pin_project must declare a struct or enum"))
    }

    let (type_, args) = match found_type {
        Some(t) => t,
        None => return Err(error!(span, "No #[pin_projectable] type found!"))
    };

    let res = handle_type(args, type_, found_pinned_drop.clone());
    res
}

pub(super) fn attribute(args: TokenStream, input: TokenStream) -> Result<TokenStream> {
    let span = span!(input);
    let item = syn::parse2(input)?;
    match &item {
        Item::Struct(_) | Item::Enum(_) => handle_type(args, item, None),
        _ => Err(error!(span, "may only be used on structs or enums")),
    }
}

fn ensure_not_packed(attrs: &[Attribute]) -> Result<()> {
    for attr in attrs {
        if let Ok(meta) = attr.parse_meta() {
            if let Meta::List(l) = meta {
                if l.ident == "repr" {
                    for repr in l.nested.iter() {
                        if let NestedMeta::Meta(Meta::Word(w)) = repr {
                            if w == "packed" {
                                return Err(error!(w, "pin_projectable may not be used on #[repr(packed)] types"))
                            }
                        }
                    }
                }
            }
        }
    }
    return Ok(())
}

/// Makes the generics of projected type from the reference of the original generics.
fn proj_generics(generics: &Generics) -> Generics {
    let mut generics = generics.clone();
    generics.params.insert(0, syn::parse_quote!('__a));
    generics
}

struct ImplDrop {
    generics: Generics,
    span: Span,
    pinned_drop: Option<ItemFn>
}

impl ImplDrop {
    /// Parses attribute arguments.
    fn new(args: TokenStream, generics: Generics, pinned_drop: Option<ItemFn>) -> Result<Self> {
        Ok(ImplDrop { generics, span: args.span(), pinned_drop })
    }

    fn build(self, ident: &Ident)-> TokenStream {
        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();

        match self.pinned_drop {
            Some(fn_) => {
                let fn_name = fn_.ident.clone();
                quote_spanned! { self.span =>
                    impl #impl_generics ::core::ops::Drop for #ident #ty_generics #where_clause {
                        fn drop(&mut self) {
                            // Declare the #[pinned_drop] function *inside* our drop function
                            // This guarantees that it's impossible for any other user code
                            // to call it
                            #fn_
                            // Safety - we're in 'drop', so we know that 'self' will
                            // never move again
                            let pinned_self = unsafe { ::core::pin::Pin::new_unchecked(self) };
                            // 'pinned_drop' is a free function - if it were part of a trait impl,
                            // it would be possible for user code to call it by directly invoking
                            // the trait.
                            // Therefore, we enforce a return type of '()' by explicitly
                            // assigning it to a temporary.
                            let _: () = #fn_name(pinned_self);
                        }
                    }
                }
            },
            None => {
                quote_spanned! { self.span =>
                    impl #impl_generics ::core::ops::Drop for #ident #ty_generics #where_clause {
                        fn drop(&mut self) {
                            // Do nothing. The precense of this Drop
                            // impl ensures that the user can't provide one of their own
                        }
                    }
                }
            }
        }
    }
}

// =================================================================================================
// conditional Unpin implementation

struct ImplUnpin {
    generics: Generics,
    span: Span,
    auto: bool
}

impl ImplUnpin {
    /// Parses attribute arguments.
    fn new(args: TokenStream, generics: &Generics) -> Result<Self> {
        let mut generics = generics.clone();
        generics.make_where_clause();

        match &*args.to_string() {
            "" => Ok(Self { generics: generics.clone(), span: args.span(), auto: true }),
            "unsafe_Unpin" => Ok(Self { generics: generics.clone(), span: args.span(), auto: false }),
            _ => Err(error!(args, "an invalid argument was passed")),
        }
    }

    fn push(&mut self, ty: &Type) {
        // We only add bounds for automatically generated impls
        if self.auto {
            self.generics
                .make_where_clause()
                .predicates
                .push(syn::parse_quote!(#ty: ::core::marker::Unpin));
        }
    }

    /// Creates `Unpin` implementation.
    fn build(self, ident: &Ident) -> TokenStream {
        let (impl_generics, ty_generics, where_clause) = self.generics.split_for_impl();
        let mut where_clause = where_clause.unwrap().clone(); // Created in 'new'
        println!("where_clause: {:?}", where_clause);

        let res = if self.auto {
            quote_spanned! { self.span =>
                impl #impl_generics ::core::marker::Unpin for #ident #ty_generics #where_clause {}
            }
        } else {
            println!("Quoting: {:?}", impl_generics);
            where_clause.predicates.push(syn::parse_quote!(#ident #ty_generics: ::pin_project::UnsafeUnpin));

            let workaround_ident = Ident::new(&("__rust_workaround_".to_string() + &ident.to_string()), Span::call_site());

            quote_spanned! { self.span =>
                impl #impl_generics ::core::marker::Unpin for #ident #ty_generics #where_clause {}
                fn #workaround_ident #impl_generics () #where_clause {}
            }


        };
        println!("Res: {}", res);
        res
    }
}
