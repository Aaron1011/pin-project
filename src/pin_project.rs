use proc_macro2::{Group, Ident, TokenStream, TokenTree};
use quote::ToTokens;
use syn::{
    parse::{Parse, ParseStream},
    punctuated::Punctuated,
    visit_mut::VisitMut,
    *,
};

use crate::project::Dummy;

pub(super) fn attribute(args: TokenStream, input: TokenStream) -> TokenStream {
    syn::parse2(input)
        .and_then(|mut item| {
            syn::parse2(args).map(|args: Args| {
                // TODO: Integrate into `replace_item_fn`?.
                Dummy.visit_item_fn_mut(&mut item);
                replace_item_fn(&args.0, &mut item);
                item.into_token_stream()
            })
        })
        .unwrap_or_else(|e| e.to_compile_error())
}

fn replace_item_fn(args: &[Ident], ItemFn { decl, block, .. }: &mut ItemFn) {
    decl.inputs.iter_mut().for_each(|input| match input {
        FnArg::Captured(ArgCaptured {
            pat: Pat::Ident(pat @ PatIdent { subpat: None, .. }),
            ..
        }) if args.contains(&pat.ident) => {
            let mut local = Local {
                attrs: Vec::new(),
                let_token: token::Let::default(),
                pats: Punctuated::new(),
                ty: None,
                init: None,
                semi_token: token::Semi::default(),
            };
            let (local_pat, init) = if pat.ident == "self" {
                ReplaceSelf.visit_block_mut(block);
                let mut local_pat = pat.clone();
                prepend_underscores_to_self(&mut local_pat.ident);
                (local_pat, syn::parse_quote!(self.project()))
            } else {
                let ident = &pat.ident;
                (pat.clone(), syn::parse_quote!(#ident.project()))
            };
            local.pats.push(Pat::Ident(local_pat));
            local.init = Some((token::Eq::default(), init));
            block.stmts.insert(0, Stmt::Local(local));

            if pat.by_ref.is_none() {
                pat.mutability = None;
            }
            pat.by_ref = None;
        }
        _ => {}
    })
}

struct Args(Vec<Ident>);

impl Parse for Args {
    fn parse(input: ParseStream<'_>) -> syn::Result<Self> {
        let mut args = Vec::new();
        let mut first = true;
        while !input.is_empty() {
            if first {
                first = false;
            } else {
                let _: Token![,] = input.parse()?;
                if input.is_empty() {
                    break;
                }
            }

            let ident = if input.peek(Token![self]) {
                let t: Token![self] = input.parse()?;
                Ident::new("self", t.span)
            } else {
                input.parse()?
            };
            if args.contains(&ident) {
                // TODO: error
            } else {
                args.push(ident);
            }
        }
        Ok(Self(args))
    }
}

// https://github.com/dtolnay/no-panic/blob/master/src/lib.rs

struct ReplaceSelf;

impl VisitMut for ReplaceSelf {
    fn visit_expr_path_mut(&mut self, i: &mut ExprPath) {
        if i.qself.is_none() && i.path.is_ident("self") {
            prepend_underscores_to_self(&mut i.path.segments[0].ident);
        }
    }

    fn visit_macro_mut(&mut self, i: &mut Macro) {
        // We can't tell in general whether `self` inside a macro invocation
        // refers to the self in the argument list or a different self
        // introduced within the macro. Heuristic: if the macro input contains
        // `fn`, then `self` is more likely to refer to something other than the
        // outer function's self argument.
        if !contains_fn(i.tts.clone()) {
            i.tts = fold_token_stream(i.tts.clone());
        }
    }

    fn visit_item_mut(&mut self, _i: &mut Item) {
        // Do nothing, as `self` now means something else.
    }
}

fn contains_fn(tts: TokenStream) -> bool {
    tts.into_iter().any(|tt| match tt {
        TokenTree::Ident(ident) => ident == "fn",
        TokenTree::Group(group) => contains_fn(group.stream()),
        _ => false,
    })
}

fn fold_token_stream(tts: TokenStream) -> TokenStream {
    tts.into_iter()
        .map(|tt| match tt {
            TokenTree::Ident(mut ident) => {
                prepend_underscores_to_self(&mut ident);
                TokenTree::Ident(ident)
            }
            TokenTree::Group(group) => {
                let content = fold_token_stream(group.stream());
                TokenTree::Group(Group::new(group.delimiter(), content))
            }
            other => other,
        })
        .collect()
}

fn prepend_underscores_to_self(ident: &mut Ident) {
    if ident == "self" {
        *ident = Ident::new("__self", ident.span());
    }
}
