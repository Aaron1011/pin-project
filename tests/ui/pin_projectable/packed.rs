// compile-fail

#![deny(warnings)]


use pin_project::pin_projectable;

#[pin_projectable]
#[repr(packed, C)] //~ ERROR may not be used on #[repr(packed)] type
struct Foo {
    #[pin]
    field: u8,
}

// Test putting 'repr' before the 'pin_projectable' attribute
#[repr(packed, C)] //~ ERROR may not be used on #[repr(packed)] type
#[pin_projectable]
struct Foo2 {
    #[pin]
    field: u8,
}

fn main() {}
