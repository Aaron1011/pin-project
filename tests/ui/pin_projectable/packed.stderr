error: pin_projectable may not be used on #[repr(packed)] types
 --> $DIR/packed.rs:8:8
  |
8 | #[repr(packed, C)] //~ ERROR may not be used on #[repr(packed)] type
  |        ^^^^^^

error: pin_projectable may not be used on #[repr(packed)] types
  --> $DIR/packed.rs:15:8
   |
15 | #[repr(packed, C)] //~ ERROR may not be used on #[repr(packed)] type
   |        ^^^^^^

error: pin_projectable may not be used on #[repr(packed)] types
  --> $DIR/packed.rs:23:8
   |
23 | #[repr(packed, C)] //~ ERROR may not be used on #[repr(packed)] type
   |        ^^^^^^

error: pin_projectable may not be used on #[repr(packed)] types
  --> $DIR/packed.rs:28:8
   |
28 | #[repr(packed, C)] //~ ERROR may not be used on #[repr(packed)] type
   |        ^^^^^^

error: aborting due to 4 previous errors

